File: bending-api/src/main/java/pw/dotdash/bending/api/ability/AbilityType.java
package pw.dotdash.bending.api.ability;

import ninja.leaping.configurate.ConfigurationNode;
import org.spongepowered.api.CatalogType;
import org.spongepowered.api.Sponge;
import org.spongepowered.api.text.Text;
import org.spongepowered.api.util.ResettableBuilder;
import pw.dotdash.bending.api.element.Element;

import java.util.Collection;
import java.util.Optional;
import java.util.function.Function;

/**
 * A type of ability.
 */
public interface AbilityType extends CatalogType {

    /**
     * Creates a new {@link Builder} to build a {@link AbilityType}.
     *
     * @return The builder
     */
    static Builder builder() {
        return Sponge.getRegistry().createBuilder(Builder.class);
    }

    /**
     * Attempts to retrieve the {@link AbilityType} based on
     * the string id given.
     *
     * @param id The case insensitive string id
     * @return The found ability type, if available
     */
    static Optional<AbilityType> get(String id) {
        return Sponge.getRegistry().getType(AbilityType.class, id);
    }

    /**
     * Gets a collection of all available {@link AbilityType}s.
     *
     * @return A collection of all known ability types
     */
    static Collection<AbilityType> getAll() {
        return Sponge.getRegistry().getAllOf(AbilityType.class);
    }

    /**
     * Gets the {@link Element} this {@link AbilityType} is associated with.
     *
     * @return The associated element
     */
    Element getElement();

    /**
     * Gets all possible ways of executing the ability.
     *
     * @return All ability execution types
     */
    Collection<AbilityExecutionType> getExecutionTypes();

    /**
     * Gets instructions on how to execute the ability.
     *
     * @return The instructions, if available
     */
    Optional<Text> getInstructions();

    /**
     * Gets the description of what the ability does.
     *
     * @return The description, if available
     */
    Optional<Text> getDescription();

    Text show();

    /**
     * Loads an ability configuration by the given {@link ConfigurationNode}.
     *
     * @param node The node used for loading
     * @return The ability configuration, if available
     */
    Optional<Ability> load(ConfigurationNode node);

    /**
     * Represents a builder to create a {@link AbilityType}.
     */
    interface Builder extends ResettableBuilder<AbilityType, Builder> {

        /**
         * Sets the string id to be used for {@link CatalogType#getId()}.
         *
         * @param id The string id
         * @return This builder, for chaining
         */
        Builder id(String id);

        /**
         * Sets the human readable name to be used for {@link CatalogType#getName()}.
         *
         * @param name The human readable name
         * @return This builder, for chaining
         */
        Builder name(String name);

        /**
         * Sets the associated element.
         *
         * @param element The associated element
         * @return This builder, for chaining
         */
        Builder element(Element element);

        Builder executionTypes(Collection<AbilityExecutionType> executionTypes);

        Builder executionTypes(AbilityExecutionType... executionTypes);

        Builder instructions(Text instructions);

        Builder description(Text description);

        Builder configLoader(Function<ConfigurationNode, Optional<Ability>> configLoader);

        default Builder loader(Function<ConfigurationNode, Ability> configLoader) {
            return this.configLoader(node -> Optional.of(configLoader.apply(node)));
        }

        /**
         * Builds a new {@link AbilityType}, provided that the
         * {@link #id(String)}, {@link #name(String)},
         * {@link #element(Element)}, {@link #executionTypes(Collection)},
         * and {@link #configLoader(Function)} are set.
         *
         * @return The generated ability type
         */
        AbilityType build();
    }
}

File: src/main/kotlin/arvenwood/bending/api/ability/AbilityType.kt
package arvenwood.bending.api.ability

import arvenwood.bending.api.element.Element
import ninja.leaping.configurate.ConfigurationNode
import org.spongepowered.api.CatalogType
import org.spongepowered.api.Sponge
import org.spongepowered.api.text.Text
import org.spongepowered.api.text.action.TextAction
import org.spongepowered.api.text.action.TextActions
import org.spongepowered.api.text.format.TextColors
import org.spongepowered.api.util.ResettableBuilder
import kotlin.coroutines.CoroutineContext

interface AbilityType<out T : Ability<T>> : CatalogType, CoroutineContext.Element {

    companion object : CoroutineContext.Key<AbilityType<Ability<*>>> {

        @JvmStatic
        fun <T : Ability<T>> builder(): Builder<T> =
            @Suppress("UNCHECKED_CAST")
            (Sponge.getRegistry().createBuilder(Builder::class.java) as Builder<T>)

        operator fun get(id: String): AbilityType<*>? =
            Sponge.getRegistry().getType(AbilityType::class.java, id).orElse(null)

        val all: Collection<AbilityType<*>>
            get() = Sponge.getRegistry().getAllOf(AbilityType::class.java)
    }

    override fun getId(): String

    override fun getName(): String

    val element: Element

    val executionTypes: Set<AbilityExecutionType>

    val instructions: Text

    val description: Text

    fun load(node: ConfigurationNode): T

    fun show(): Text =
        Text.builder(this.name)
            .color(this.element.color)
            .onHover(TextActions.showText(Text.of(this.element.color, this.description, "\n\nInstructions:\n", TextColors.GOLD, this.instructions)))
            .build()

    override val key: CoroutineContext.Key<*> get() = AbilityType

    interface Builder<out T : Ability<T>> : ResettableBuilder<AbilityType<@UnsafeVariance T>, Builder<@UnsafeVariance T>> {

        fun id(id: String): Builder<T>

        fun name(name: String): Builder<T>

        fun element(element: Element): Builder<T>

        fun executionTypes(executionTypes: Set<AbilityExecutionType>): Builder<T>

        fun executionTypes(vararg executionTypes: AbilityExecutionType): Builder<T> =
            executionTypes(executionTypes.toSet())

        fun instructions(instructions: Text): Builder<T>

        fun description(description: Text): Builder<T>

        fun loader(loader: (ConfigurationNode) -> @UnsafeVariance T): Builder<T>

        fun build(): AbilityType<T>
    }
}